Add smoke tests for Interpreter Dropdown

sharon-wang opened the pull request:
## Description

- Addresses https://github.com/posit-dev/positron/issues/4124
- Adds new test cases
    - Python interpreter starts and shows running
    - Python interpreter restarts and shows running
    - R interpreter starts and shows running
    - R interpreter stops and shows inactive

### QA Notes

- please add TestRail IDs 😄
- please mark tests for #pr CI runs as appropriate
- `StartInterpreter` was refactored significantly (and renamed to `PositronInterpreterDropdown`), but was only previously used by `new-project.test.ts`

sharon-wang committed:
rename `StartInterpreter` to `PositronInterpreterDropdown` for consistency

sharon-wang committed:
update `new-project.test.ts` to use renamed interpreter dropdown class

sharon-wang committed:
add interpreter dropdown initial test

includes new util methods and some refactoring for the positronInterpreterDropdown class

sharon-wang committed:
add test to start R interpreter from dropdown

sharon-wang committed:
add interpreter restart smoke test

sharon-wang committed:
add test for stopping interpreter from dropdown

sharon-wang commented:
Full test suite passed 🎉, although the push to slack failed:

```
Traceback (most recent call last):
  File "/source/app/script.py", line 109, in <module>
    main()
  File "/source/app/script.py", line 91, in main
    slack_utils.send_slack_msg(
  File "/source/app/utils/slack_utils.py", line 19, in send_slack_msg
    client.chat_postMessage(
  File "/usr/local/lib/python3.10/site-packages/slack/web/client.py", line 1077, in chat_postMessage
    return self.api_call("chat.postMessage", json=kwargs)
  File "/usr/local/lib/python3.10/site-packages/slack/web/base_client.py", line [15](https://github.com/posit-dev/positron/actions/runs/10290770371/job/28481507074#step:24:16)0, in api_call
    return self._sync_send(api_url=api_url, req_args=req_args)
  File "/usr/local/lib/python3.10/site-packages/slack/web/base_client.py", line 241, in _sync_send
    return self._urllib_api_call(
  File "/usr/local/lib/python3.10/site-packages/slack/web/base_client.py", line 378, in _urllib_api_call
    ).validate()
  File "/usr/local/lib/python3.10/site-packages/slack/web/slack_response.py", line [19](https://github.com/posit-dev/positron/actions/runs/10290770371/job/28481507074#step:24:20)4, in validate
    raise e.SlackApiError(message=msg, response=self)
slack.errors.SlackApiError: The request to the Slack API failed.
The server responded with: {'ok': False, 'error': 'not_in_channel'}
```

testlabauto commented:
> Full test suite passed 🎉, although the push to slack failed:
> 
> ```
> Traceback (most recent call last):
>   File "/source/app/script.py", line 109, in <module>
>     main()
>   File "/source/app/script.py", line 91, in main
>     slack_utils.send_slack_msg(
>   File "/source/app/utils/slack_utils.py", line 19, in send_slack_msg
>     client.chat_postMessage(
>   File "/usr/local/lib/python3.10/site-packages/slack/web/client.py", line 1077, in chat_postMessage
>     return self.api_call("chat.postMessage", json=kwargs)
>   File "/usr/local/lib/python3.10/site-packages/slack/web/base_client.py", line [15](https://github.com/posit-dev/positron/actions/runs/10290770371/job/28481507074#step:24:16)0, in api_call
>     return self._sync_send(api_url=api_url, req_args=req_args)
>   File "/usr/local/lib/python3.10/site-packages/slack/web/base_client.py", line 241, in _sync_send
>     return self._urllib_api_call(
>   File "/usr/local/lib/python3.10/site-packages/slack/web/base_client.py", line 378, in _urllib_api_call
>     ).validate()
>   File "/usr/local/lib/python3.10/site-packages/slack/web/slack_response.py", line [19](https://github.com/posit-dev/positron/actions/runs/10290770371/job/28481507074#step:24:20)4, in validate
>     raise e.SlackApiError(message=msg, response=self)
> slack.errors.SlackApiError: The request to the Slack API failed.
> The server responded with: {'ok': False, 'error': 'not_in_channel'}
> ```

Oops.  I forgot to invite the app to the new channel.  Done now!

testlabauto reviewed (dismissed):
Nicely done!  Appreciate the excellent refactor of positronStartInterpreter.ts!

testlabauto committed:
Added testrail refs

sharon-wang committed:
fix comment typos and log wording

sharon-wang committed:
disable auto interpreter startup for interpreter dropdown tests

The setting `'positron.interpreters.automaticStartup'` is disabled before the interpreter dropdown tests are run. After the tests are run, the changed setting is cleared.

The automatic startup of an interpreter was bulldozing the starting of the interpreter selected in the dropdown. This was causing the first test to fail.

sharon-wang commented on test/smoke/src/fixtures/positronConsoleFixtures.ts:
This file has been replaced by `test/automation/src/positron/fixtures/positronUserSettingsFixtures.ts`

sharon-wang commented on test/smoke/src/areas/positron/top-action-bar/interpreter-dropdown.test.ts:
```diff
@@ -0,0 +1,187 @@
+ */
+export function setup(logger: Logger) {
+	describe('Interpreter Dropdown in Top Action Bar', () => {
+		/**
+		 * NOTE: This describe block MUST RUN FIRST, before any of the other tests in this suite.
+		 * This unfortunately cannot be run in the 'before' block of the following describe block,
+		 * because the app restart interferes with the timing of the tests (the app is not ready
+		 * when the tests start).
+		 */
+		describe('Interpreter Dropdown in Top Action Bar - setup', async () => {
+			// Shared before/after handling
+			installAllHandlers(logger);
+
+			let app: Application;
+			let userSettings: PositronUserSettingsFixtures;
+
+			before(async function () {
+				app = this.app as Application;
+				userSettings = new PositronUserSettingsFixtures(app);
+			});
+
+			it('Disable automatic startup of interpreters', async function () {
+				// Disable automatic startup of interpreters to prevent the auto-started interpreter from
+				// bulldozing the interpreter dropdown selection
+				await userSettings.setUserSetting([
+					'positron.interpreters.automaticStartup',
+					'false',
+				]);
+			});
+		});
```
Disables the auto interpreter setup and the app will be restarted between this describe block and the actual test describe block

sharon-wang replied:
this has been removed

sharon-wang commented on test/smoke/src/areas/positron/top-action-bar/interpreter-dropdown.test.ts:
```diff
@@ -0,0 +1,187 @@
+			});
+		});
+
+		/**
+		 * NOTE: This describe block MUST RUN LAST, after any of the other tests in this suite.
+		 * This unfortunately cannot be run in the 'after' block of the previous describe block,
+		 * as Positron closes before or while the after block is running.
+		 */
+		describe('Interpreter Dropdown in Top Action Bar - cleanup', async () => {
+			// Shared before/after handling
+			installAllHandlers(logger);
+
+			let app: Application;
+			let userSettings: PositronUserSettingsFixtures;
+
+			before(async function () {
+				app = this.app as Application;
+				userSettings = new PositronUserSettingsFixtures(app);
+			});
+
+			it('Unset user settings', async function () {
+				// Unset the user setting to re-enable automatic startup of interpreters
+				await userSettings.unsetUserSettings();
+			});
+		});
+	});
```
Clears the user settings after the tests have run, to restore the original behaviour

sharon-wang replied:
this has been removed

sharon-wang reviewed (commented):
I've add "setup" and "cleanup" describe blocks with single "tests" to set and unset the auto interpreter startup user setting. I wasn't able to use the `before` and `after` blocks in the existing `describe`, as the app restart timing interferes with the tests (more info in the comments).

Full test suite in progress: https://github.com/posit-dev/positron/actions/runs/10309907612

Hope to see this pass on Windows too! 🤞

sharon-wang committed:
add `waitForReadyOrNoInterpreter` check to `positronConsole`

Remove the user settings changes from the interpreter dropdown test and just wait for the console to be ready, whether with an active interpreter or not.

sharon-wang committed:
use a hybrid between user settings and shutting down interpreter

I can simulate auto-starting an interpreter by adding:

```
			// console.log('===== "auto-starting" interpreter: Python 3.12.4');
			// await positronConsole.selectInterpreter('Python', 'Python 3.12.4');
			// console.log('===== artificial wait...');;
			// await app.code.wait(1000);
```

to the `before` block before the "ensure no interpreters running" handling

sharon-wang commented on test/smoke/src/areas/positron/top-action-bar/interpreter-dropdown.test.ts:
```diff
@@ -0,0 +1,137 @@
+
+		beforeEach(async function () {
+			await positronConsole.waitForReadyOrNoInterpreter();
+		});
```
this is removed now

sharon-wang committed:
increase select interpreter timeout

jonvanausdeln reviewed (approved):
LGTM! New timeout works on my machine!  Hopefully the CI pass works!

sharon-wang commented:
CI is green! ✅ 🎉
